src/main/java/tech/picnic/jolo/ObjectGraph.java

```diff
@@ -0,0 +1,139 @@
+package tech.picnic.jolo;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static tech.picnic.jolo.Util.validate;
+
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.MultimapBuilder;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Table;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.StringJoiner;
+import javax.annotation.Nullable;
+import org.jooq.Record;
+
+/**
+ * Mutable graph of objects loaded by {@link Entity entities}. Objects are linked by {@link IdPair
+ * ID pairs} loaded by {@link Relation relations}.
+ *
+ * @apiNote Access is not synchronized.
+ */
+final class ObjectGraph {
+  private final Table<Entity<?, ?>, Long, Object> entityAndIdToObject;
+  private final SetMultimap<Relation<?, ?>, IdPair> relationToLinks;
+
+  ObjectGraph() {
+    entityAndIdToObject = HashBasedTable.create();
+    relationToLinks = MultimapBuilder.hashKeys().linkedHashSetValues().build();
+  }
+
+  /**
+   * Add an object loaded by the given {@link Entity entity} with the given ID. If an object of the
+   * same entity and with the same ID already exists, the given object is ignored.
+   */
+  <E> void add(Entity<? extends E, ?> entity, long id, E object) {
+    requireNonNull(object);
+    if (!entityAndIdToObject.contains(entity, id)) {
+      entityAndIdToObject.put(entity, id, object);
+    }
+  }
+
+  /** Add links loaded by the given {@link Relation relation}. */
+  <L, R> void add(Relation<L, R> relation, Set<IdPair> links) {
+    relationToLinks.putAll(relation, links);
+  }
+
+  /**
+   * Merges the other graph into this one.
+   *
+   * @param other The other object graph to merge with.
+   * @apiNote This operation is associative but not commutative. Existing objects loaded by the same
+   *     entity and with the same ID take precedence.
+   */
+  void merge(ObjectGraph other) {
+    requireNonNull(other);
+    for (var cell : other.entityAndIdToObject.cellSet()) {
+      add(cell.getRowKey(), cell.getColumnKey(), cell.getValue());
+    }
```

Comment:
`entityAndIdToObject.putAll(other.entityAndIdToObject)` would override existing cells but we want to give precedence to existing ones in case of conflicts.